File: python/puzzle02.py
from pathlib import Path
from collections import Counter


with open(Path(__file__).parent.parent / "data" / "puzzle02.txt", "r") as f:
    raw_data = f.read().splitlines()


def check_password_entry(entry: str) -> bool:
    policy, password = entry.split(': ')
    policy_char_range_string, policy_char = policy.split(' ')
    policy_char_range_list = [int(ele) for ele in policy_char_range_string.split('-')]
    policy_char_range = range(policy_char_range_list[0], policy_char_range_list[1] + 1)

    return Counter(password)[policy_char] in policy_char_range


def check_password_entry_complex(entry: str) -> bool:
    password = entry.split(':')[1].strip()
    policy = entry.split(':')[0]
    policy_char = policy.split(' ')[1]
    policy_char_range_list = list(map(int, policy.split(' ')[0].split('-')))
    first_occurrence = policy_char_range_list[0] - 1
    second_occurrence = policy_char_range_list[1] - 1

    return (password[first_occurrence] == policy_char) != (password[second_occurrence] == policy_char)


print(f'Solution for A is {sum(check_password_entry(password_entry) for password_entry in raw_data)}')  # 546
print(f'Solution for B is {sum(check_password_entry_complex(password_entry) for password_entry in raw_data)}')  # 275

File: julia/puzzle02.jl
import DataStructures

raw_data = readlines("data/puzzle02.txt")


function check_password_entry_a(entry::String)::Bool
    policy, password = split(entry, ": ")
    policy_char_range_string, policy_char = split(policy, ' ')
    policy_char_range_arr = collect(parse(Int, ele) for ele in split(policy_char_range_string, "-"))
    policy_char_range = UnitRange(policy_char_range_arr[1], policy_char_range_arr[2])
    return DataStructures.counter(password)[policy_char[1]] in policy_char_range
end


function check_password_entry_b(entry::String)::Bool
    policy, password = split(entry, ": ")
    policy_char_range_string, policy_char = split(policy, ' ')
    policy_char_range = collect(parse(Int, ele) for ele in split(policy_char_range_string, "-"))
    return (password[policy_char_range[1]] == policy_char[1]) != (password[policy_char_range[2]] == policy_char[1])
end

println("Result for A is $(sum(check_password_entry_a(ele) for ele in raw_data))")
println("Result for B is $(sum(check_password_entry_b(ele) for ele in raw_data))")
